### ISSUE TITLE
Write Gadget 2 snapshots

### ISSUE THREAD

**Issue by TiagoBsCastro** [pynbody/pynbody#366]
I've been trying to rewrite a format 2 gadget snapshot that has been written in parallel (and so was splited in N "sub-snapshots") in one single snapshot. I tried the following:

snap=pynbody.load("snapshot_inhomo_009") 
snap.write(filename="snapshot_inhomo_009.g2",fmt=pynbody.gadget.GadgetSnap)

but it returns me the message:

```
---------------------------------------------------------------------------
ValueError Traceback (most recent call last)
<ipython-input-10-4a0a5ab9fa39> in <module>()
----> 1 snap.write(filename="snapshot_inhomo_009",fmt=pynbody.gadget.GadgetSnap)

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/__init__.pyc in write(self, fmt, filename, **kwargs)
  984 self._write(self, filename, **kwargs)
  985 else:
--> 986 fmt._write(self, filename, **kwargs)
  987 
  988 def write_array(self, array_name, fam=None, overwrite=False, **kwargs):

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/gadget.pyc in _write(self, filename)
  1081 # Call _write_array for every array.
  1082 for x in all_keys:
-> 1083 GadgetSnap._write_array(self, x, filename=filename)
  1084 
  1085 @staticmethod

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/gadget.pyc in _write_array(self, array_name, fam, filename)
  1117 for f in self._files[:-2]:
  1118 f.add_file_block(array_name, per_file, ashape[
-> 1119 1], dtype=self[array_name].dtype, p_types=p_types)
  1120 self._files[-1].add_file_block(
  1121 array_name, npart - (nfiles - 1) * per_file, ashape[1])

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/gadget.pyc in add_file_block(self, name, blocksize, partlen, dtype, p_types)
  591 block.start = lb.start + lb.length + 6 * \
  592 4 # For the block header, and footer of the previous block
--> 593 if p_types == -1:
  594 block.p_types = np.ones(N_TYPE, bool)
  595 else:

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Am I doing something wrong?

**Comment by apontzen** [comment]
I can't reproduce this error with other multi-file gadget snapshots - could you post the actual files you are using?

**Comment by TiagoBsCastro** [comment]
Many thanks, @apontzen 

Here it goes, although I have to say it is big...
https://www.dropbox.com/s/7x50jkrhqerll7i/snaphots.tar.gz?dl=0

**Comment by apontzen** [comment]
Very weird behaviour indeed! I think it might be something about having mass in the header but I'm not sure as I don't really use gadget files myself.

@TiagoBsCastro could you try checking out branch issue-366 to see if it fixes it for you
@sbird as the original author of the gadget code, could you comment on this? I am feeling around in the dark a bit in this attempted fix.

**Comment by sbird** [comment]
Hi Andrew - what you did looks plausible, but I'll have to get back to you in a few days on this.

**Comment by apontzen** [comment]
@TiagoBsCastro any update - did you try the fix?

**Comment by TiagoBsCastro** [comment]
Hi @apontzen, sorry I haven't had time to check your suggestion until today. Trying the same thing it returns me:

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-3-4e4a6f8fb748> in <module>()
----> 1 snap.write(filename="/home/tcastro/Documents/snapshot_inhomo_009.g2",fmt=pynbody.gadget.GadgetSnap)

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/__init__.pyc in write(self, fmt, filename, **kwargs)
    984             self._write(self, filename, **kwargs)
    985         else:
--> 986             fmt._write(self, filename, **kwargs)
    987 
    988     def write_array(self, array_name, fam=None, overwrite=False, **kwargs):

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/gadget.pyc in _write(self, filename)
   1081             # Call _write_array for every array.
   1082             for x in all_keys:
-> 1083                 GadgetSnap._write_array(self, x, filename=filename)
   1084 
   1085     @staticmethod

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/gadget.pyc in _write_array(self, array_name, fam, filename)
   1158                                 data = np.asanyarray(data, dtype=np.float32)
   1159                             self._files[i].write_block(g_name, gfam, data[
-> 1160                                                        s:(s + f_parts[i])], filename=ffile)
   1161                         s += f_parts[i]
   1162 

/usr/local/lib/python2.7/dist-packages/pynbody/snapshot/gadget.pyc in write_block(self, name, p_type, big_data, filename)
    531         parts = self.get_block_parts(name, p_type)
    532         p_start = self.get_start_part(name, p_type)
--> 533         MinType = np.ravel(np.where(cur_block.p_types * self.header.npart))[0]
    534         MaxType = np.ravel(np.where(cur_block.p_types * self.header.npart))[-1]
    535         # Have we been given the right number of particles?

IndexError: index 0 is out of bounds for axis 0 with size 0

**Comment by sbird** [comment]
From ff609d91f90099cc6a4960201e62d8cbaa6ceb75 Mon Sep 17 00:00:00 2001
From: Simeon Bird <sbird4@jhu.edu>
Date: Fri, 31 Mar 2017 22:30:17 -0700
Subject: [PATCH] Partially fix issue #366: Do not try to write blocks when
 they contain no particles

---
 pynbody/snapshot/gadget.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/pynbody/snapshot/gadget.py b/pynbody/snapshot/gadget.py
index 0de829f..311d110 100644
--- a/pynbody/snapshot/gadget.py
+++ b/pynbody/snapshot/gadget.py
@@ -1131,6 +1131,8 @@ class GadgetSnap(SimSnap):
                     f_parts = [f.get_block_parts(
                         g_name, gfam) for f in self._files]
                     for i in np.arange(0, nfiles):
+                        if f_parts[i] == 0:
+                            continue
                         # Set up filename
                         if filename != None:
                             ffile = filename + "." + str(i)
-- 
2.9.3

**Comment by sbird** [comment]
Try that patch

**Comment by sbird** [comment]
By the way, I also confirm Andrew's patch

**Comment by apontzen** [comment]
@TiagoBsCastro I have applied @sbird's additional patch to branch issue-366, could you pull and try again? Thanks.

**Comment by TiagoBsCastro** [comment]
Hi all, many thanks. I've tried the new version and it worked.

**Comment by apontzen** [comment]
Merged into master - thanks.
